hpcgap/lib/ffeconway.gi

```diff
@@ -1196,225 +1196,93 @@ end);
+    if x = baby[i+1] then
+      return i;
+    fi;
+  od;
+  giant := baby[rr]*base;
+  ord := [0..rr-1];
+  SortParallel(baby, ord);
+  t := x;
+  for j in [1..QuoInt(r, rr)+1] do
+    t := t*giant;
+    pos := PositionSet(baby, t);
+    if IsInt(pos) then
+      return (ord[pos] - j * rr) mod r;
+    fi;
+  od;
+  return fail;
 end;
 
+# recursive method, m can be order of base or its factorization
+# Let r be the largest prime factor of m, then we use
+#     base^e = x with e = a + b*r where 0 <= a < r and
+# 0 <= b < m/r, and compute a with DLogShanks and b by
+# recursion.
+FFECONWAY.DLog:= function(base, x, m...)
+  local r, mm, mp, a, b;
+  if Length(m) = 0 then
+    m := Order(base);
+  else
+    m := m[1];
+  fi;
+  if not IsList(m) then
+    m := Factors(m);
+  fi;
+  if Length(m) = 1 then
+    return FFECONWAY.DLogShanks(base, x, m[1]);
+  fi;
+  r := m[Length(m)];
+  mm := m{[1..Length(m)-1]};
+  mp := Product(mm);
+  a := FFECONWAY.DLogShanks(base^mp, x^mp, r);
+  b := FFECONWAY.DLog(base^r, x/(base^a), mm);
+  return a + b*r;
+end;
 
-
-FFECONWAY.DoLogFFERho :=function(y,z,ord,f,q)
-    local   fact,  s,  i,  t,  d,  Q,  R,  MN,  M,  N,  rep,  k,
-            theta,  Qp,  o;
-    fact:=[];
-    s:=ord;
-    for i in f do
-        t:=s/i;
-        if IsInt(t) then
-            s:=t;
-            Add(fact,i);
-        fi;
-    od;
-
-    if Length(fact)>1 then
-        d:=ord;
-        while (d=ord) and Length(fact)>0 do
-            s:=Remove(fact);
-            t:=ord/s;
-            Q:=y^s;
-            R:=z^s;
-
-        # iterate
-            MN:=FFECONWAY.LogFFERhoIterate(Q,R,q);
-            M:=MN[1];
-            N:=MN[2];
-            rep:=GcdRepresentation(ord,s*M);
-            d:=rep[1]*ord+rep[2]*s*M;
-        od;
-
-        if d < ord then
-            k:=(rep[2]*s*N/d);
-            if Gcd(DenominatorRat(k),ord)<>1 then
-                return fail;        # can't invert (can't happen if not primitive root)
-            fi;
-            k:=k mod ord;
-            theta:= z^(ord/d);
-
-            Qp:=y/(z^k);
-            i:=FFECONWAY.DoLogFFERho(Qp,theta,d,f,q);
-            if i=fail then return i;fi; # bail out
-            o:=(k+i*(ord/d)) mod ord;
-            Assert(1,z^o=y);
-            return o;
-        fi;
-    fi;
-
-    # naive case, iterate
-    MN:=FFECONWAY.LogFFERhoIterate(y,z,q);
-    M:=MN[1];
-    N:=MN[2];
-    rep:=GcdRepresentation(ord,M);
-    d:=rep[1]*ord+rep[2]*M;
-    k:=(rep[2]*N/d);
-    if Gcd(DenominatorRat(k),ord)<>1 then
-        return fail;        # can't invert (can't happen if not primitive root)
-    fi;
-    k:=k mod ord;
-    theta:= z^(ord/d);
-    Qp:= y/(z^k);
-    i := 0; while i < d do i := i + 1;
-        if IsOne(Qp) then
-            Assert(1,z^k = y);
-            return k;
-        fi;
-        k:=(k+ord/d) mod ord;
-        Qp:=Qp/theta;
-    od;
-    # process failed (because z was not a primitive root)
+FFECONWAY.DoLogFFE:= function(x, base)
```

Comment:
This code is in no way specific to FFECONWAY, is it? I mean, StandardFF was using it, and looking at it, it seems to be that it might in fact work for `x` and `base` elements of arbitrary multiplicative groups, right?

So perhaps it would make sense to move it elsewhere, and out of this record? And possibly drop the `FFE` from the name?

It could go into a new file, `dlog.gi` or whatever. Or an existing file (that would then also remove the need of maintaining two copies of this code for GAP vs. HPC-GAP.